audio.go

```diff
@@ -65,64 +82,58 @@ func (as *AudioSystem) New(*ecs.World) {
 		// TODO: but it is when the CameraMessage is not Incremental (i.e. the changes are big)
-		al.SetListenerPosition(al.Vector{cam.X() / Width(), cam.Y() / Height(), cam.Z() * as.HeightModifier})
+		al.SetListenerPosition(al.Vector{cam.X() / Width(), cam.Y() / Height(), cam.Z() * a.HeightModifier})
 	})
 }
 
-func (as *AudioSystem) UpdateEntity(entity *ecs.Entity, dt float32) {
-	var ac *AudioComponent
-	var ok bool
-	if ac, ok = entity.ComponentFast(ac).(*AudioComponent); !ok {
-		return
-	}
+func (a *AudioSystem) Update(dt float32) {
+	for _, e := range a.entities {
+		if e.AudioComponent.player == nil {
+			f := Files.Sound(e.AudioComponent.File)
+			if f == nil {
+				log.Println("Audio file not loaded:", e.AudioComponent.File)
+				continue
+			}
 
-	if ac.player == nil {
-		f := Files.Sound(ac.File)
-		if f == nil {
-			return
+			var err error
+			e.AudioComponent.player, err = NewPlayer(f, 0, 0)
+			if err != nil {
+				log.Println("Error initializing AudioComponent:", err)
+				continue
+			}
 		}
 
-		var err error
-		ac.player, err = NewPlayer(f, 0, 0)
-		if err != nil {
-			log.Println("Error initializing AudioSystem:", err)
-			return
+		if MasterVolume != a.cachedVolume {
+			e.AudioComponent.SetVolume(e.AudioComponent.RawVolume)
 		}
-	}
-
-	if MasterVolume != as.cachedVolume {
-		ac.SetVolume(ac.RawVolume)
-	}
 
-	if ac.player.State() != Playing {
-		if ac.player.State() == Stopped {
-			if !ac.Repeat {
-				al.RewindSources(ac.player.source)
-				al.StopSources(ac.player.source)
-				entity.RemoveComponent(ac)
-				return
+		if e.AudioComponent.player.State() != Playing {
+			if e.AudioComponent.player.State() == Stopped {
+				if !e.AudioComponent.Repeat {
+					al.RewindSources(e.AudioComponent.player.source)
+					al.StopSources(e.AudioComponent.player.source)
+					// Remove it from this system, defer because we want to be sure it doesn't interfere with
+					// looping over a.entities
+					defer a.Remove(*e.BasicEntity)
+					continue
+				}
 			}
-		}
 
-		// Prepares if the track hasn't been buffered before.
-		if err := ac.player.prepare(ac.Background, 0, false); err != nil {
-			log.Println("Error initializing AudioSystem:", err)
-			return
-		}
+			// Prepares if the track hasn't been buffered before.
+			if err := e.AudioComponent.player.prepare(e.AudioComponent.Background, 0, false); err != nil {
+				log.Println("Error initializing AudioComponent:", err)
+				continue
+			}
 
-		al.PlaySources(ac.player.source)
+			al.PlaySources(e.AudioComponent.player.source)
 
-		if !ac.Background {
-			var space *SpaceComponent
-			var ok bool
-			if space, ok = entity.ComponentFast(space).(*SpaceComponent); !ok {
-				return
+			if !e.AudioComponent.Background {
+				e.AudioComponent.player.source.SetPosition(al.Vector{
+					(e.SpaceComponent.Position.X + e.SpaceComponent.Width/2) / Width(), // TODO: ensure we're using correct Width/Height()
```

Comment:
I am assuming this is the case, because (1) current implementation uses the same Width/Height functions (so even if it's broken, no need to fix it within this PR), and (2) it's not as straightforward to fix.